File: c/DFS.c
#include <stdio.h>
#include <stdbool.h>

#define MAX_VERTICES 100

bool visited[MAX_VERTICES];
int stack[MAX_VERTICES];
int top = -1; // Stack top index

// Depth-first search function
void dfs(int graph[][MAX_VERTICES], int vertex, int numVertices) {
    printf("%d ", vertex); // Print the current vertex
    visited[vertex] = true;

    // Push the current vertex onto the stack
    stack[++top] = vertex;

    // Traverse adjacent vertices
    for (int i = 0; i < numVertices; i++) {
        if (graph[vertex][i] && !visited[i]) {
            dfs(graph, i, numVertices);
        }
    }

    // Pop the current vertex from the stack
    // and print the pop order
    printf("(%d popped) ", stack[top--]);
}

int main() {
    int numVertices, startVertex;
    printf("Enter the number of vertices: ");
    scanf("%d", &numVertices);

    printf("Enter the adjacency matrix:\n");
    int graph[MAX_VERTICES][MAX_VERTICES];
    for (int i = 0; i < numVertices; i++) {
        for (int j = 0; j < numVertices; j++) {
            scanf("%d", &graph[i][j]);
        }
    }

    printf("Enter the starting vertex: ");
    scanf("%d", &startVertex);

    printf("Depth First Search Traversal:\n");
    dfs(graph, startVertex, numVertices);

    return 0;
}

File: python/DFS.py
class graph:
    def __init__(self):
        self.graph={}

    def add_edge(self,u,v):
        if u not in self.graph:
            self.graph[u]=[]
        self.graph[u].append(v)

    def dfs_util(self,vertex,visited):
        visited[vertex]=True
        print(vertex,end=' ')

        for neighbor in self.graph.get(vertex,[]):
            if neighbor in visited and not  visited[neighbor]:
                self.dfs_util(neighbor,visited)


    def dfs(self,start_vertex):
        visited={ vertex:False for vertex in self.graph}

        print("the DFS traversal is")
        self.dfs_util(start_vertex,visited)
        print()

def main():
    my_graph=graph()

    my_graph.add_edge(0,1)
    my_graph.add_edge(0,2)
    my_graph.add_edge(1,2)
    my_graph.add_edge(2,3)
    my_graph.add_edge(3,4)
    my_graph.add_edge(3,5)
    my_graph.add_edge(4,5)

    start_vertex=0

    my_graph.dfs(start_vertex)

if __name__=="__main__":
    main()